File: libs/document-processor-next/src/runtime/scope-runtime-context.ts
import { BlueNode } from '@blue-labs/language';

export type TerminationKind = 'GRACEFUL' | 'FATAL';

export class ScopeRuntimeContext {
  private readonly triggeredQueue: BlueNode[] = [];
  private readonly bridgeableEvents: BlueNode[] = [];
  private terminated = false;
  private terminationKindValue: TerminationKind | null = null;
  private terminationReasonValue: string | null = null;
  private cutOff = false;
  private triggeredLimit = -1;
  private bridgeableLimit = -1;

  constructor(private readonly path: string) {}

  scopePath(): string {
    return this.path;
  }

  enqueueTriggered(node: BlueNode): void {
    if (this.cutOff && this.triggeredLimit >= 0 && this.triggeredQueue.length >= this.triggeredLimit) {
      return;
    }
    this.triggeredQueue.push(node);
  }

  pollTriggered(): BlueNode | undefined {
    return this.triggeredQueue.shift();
  }

  peekTriggered(): BlueNode | undefined {
    return this.triggeredQueue[0];
  }

  clearTriggered(): void {
    this.triggeredQueue.length = 0;
  }

  triggeredSize(): number {
    return this.triggeredQueue.length;
  }

  triggeredIsEmpty(): boolean {
    return this.triggeredQueue.length === 0;
  }

  triggeredSnapshot(): readonly BlueNode[] {
    return [...this.triggeredQueue];
  }

  recordBridgeable(node: BlueNode): void {
    if (this.cutOff && this.bridgeableLimit >= 0 && this.bridgeableEvents.length >= this.bridgeableLimit) {
      return;
    }
    this.bridgeableEvents.push(node);
  }

  drainBridgeableEvents(): BlueNode[] {
    let drained: BlueNode[];
    if (this.cutOff && this.bridgeableLimit >= 0 && this.bridgeableLimit < this.bridgeableEvents.length) {
      drained = this.bridgeableEvents.slice(0, this.bridgeableLimit);
    } else {
      drained = [...this.bridgeableEvents];
    }
    this.bridgeableEvents.length = 0;
    return drained;
  }

  isTerminated(): boolean {
    return this.terminated;
  }

  terminationKind(): TerminationKind | null {
    return this.terminationKindValue;
  }

  terminationReason(): string | null {
    return this.terminationReasonValue;
  }

  finalizeTermination(kind: TerminationKind, reason: string | null = null): void {
    if (this.terminated) {
      return;
    }
    this.terminated = true;
    this.terminationKindValue = kind;
    this.terminationReasonValue = reason ?? null;
    this.clearTriggered();
  }

  markCutOff(): void {
    if (this.cutOff) {
      return;
    }
    this.cutOff = true;
    this.triggeredLimit = this.triggeredQueue.length;
    this.bridgeableLimit = this.bridgeableEvents.length;
  }

  isCutOff(): boolean {
    return this.cutOff;
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/ScopeRuntimeContext.java
package blue.language.processor;

import blue.language.model.Node;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.Objects;

/**
 * Per-scope runtime state tracked during processing.
 */
public final class ScopeRuntimeContext {

    private final String scopePath;
    private final Deque<Node> triggeredQueue = new ArrayDeque<>();
    private final List<Node> bridgeableEvents = new ArrayList<>();
    private boolean terminated;
    private TerminationKind terminationKind;
    private String terminationReason;
    private boolean cutOff;
    private int triggeredLimit = -1;
    private int bridgeableLimit = -1;

    public ScopeRuntimeContext(String scopePath) {
        this.scopePath = Objects.requireNonNull(scopePath, "scopePath");
    }

    public String scopePath() {
        return scopePath;
    }

    public Deque<Node> triggeredQueue() {
        return triggeredQueue;
    }

    public int triggeredSize() {
        return triggeredQueue.size();
    }

    public Node pollTriggered() {
        return triggeredQueue.pollFirst();
    }

    public void enqueueTriggered(Node node) {
        if (cutOff && triggeredLimit >= 0 && triggeredQueue.size() >= triggeredLimit) {
            return;
        }
        triggeredQueue.addLast(Objects.requireNonNull(node, "node"));
    }

    public void recordBridgeable(Node node) {
        if (cutOff && bridgeableLimit >= 0 && bridgeableEvents.size() >= bridgeableLimit) {
            return;
        }
        bridgeableEvents.add(Objects.requireNonNull(node, "node"));
    }

    public List<Node> drainBridgeableEvents() {
        List<Node> drained;
        if (cutOff && bridgeableLimit >= 0 && bridgeableLimit < bridgeableEvents.size()) {
            drained = new ArrayList<>(bridgeableEvents.subList(0, bridgeableLimit));
        } else {
            drained = new ArrayList<>(bridgeableEvents);
        }
        bridgeableEvents.clear();
        return drained;
    }

    public boolean isTerminated() {
        return terminated;
    }

    public TerminationKind terminationKind() {
        return terminationKind;
    }

    public String terminationReason() {
        return terminationReason;
    }

    public void finalizeTermination(TerminationKind kind, String reason) {
        if (terminated) {
            return;
        }
        terminated = true;
        terminationKind = Objects.requireNonNull(kind, "kind");
        terminationReason = reason;
        triggeredQueue.clear();
    }

    public void markCutOff() {
        if (cutOff) {
            return;
        }
        cutOff = true;
        triggeredLimit = triggeredQueue.size();
        bridgeableLimit = bridgeableEvents.size();
    }

    public enum TerminationKind {
        GRACEFUL,
        FATAL
    }
}
